### PR TITLE
Make height or width optional

### PR DESCRIPTION
**Pull request by eatyourgreens**
Allow either height or width to be optional in the URL. URLs like `/x300/` set the width to `-`and resize only on the height. URLs like `/300x/` set the height to `-` and resize only on width.

>resize width height
    proportionally reduces an image to the specified sizes. To reduce by only one dimension, another dimension can be specified as “-”. In case of an error, the server will return code 415 (Unsupported Media Type). Parameter values can contain variables. When used along with the rotate parameter, the rotation happens after reduction. 

https://nginx.org/en/docs/http/ngx_http_image_filter_module.html#image_filter

### CODE REVIEW

File: nginx.conf [hunk 1]
```diff
@@ -43,6 +43,11 @@ http {
             set $width $1;
             set $height $2;
         }
+ 
+        if ($uri ~ "^/([1-9][0-9]{0,2})xnone/") {
+            set $width $1;
+            set $height -;
```
Comment (eatyourgreens) [inline nginx.conf]: Should this be
```suggestion
            set $height '-';
```

I'm not 100% clear on nginx config syntax.

**Comment by goplayoutside3** [comment]
In our conversation in https://github.com/zooniverse/front-end-monorepo/issues/2554, I mistyped that I think the solution should be translating the url in thumbnailer.

I think there's solution in FEM's Markdownz for handling 'none' max-height rather than modifying the thumbnailer service.

Given that FEM Markdownz needs to be refactored due to deprecation of `remark-react`, I recommend this PR is closed in favor of a PR to FEM instead. For instance, we might not actually need max-height of 'none' at all.

File: nginx.conf (hunk 1 above)
Comment (camallen) [reply nginx.conf]: i've tested this and it seems to work with "-" and '-'

**Review by camallen — COMMENTED** [review]
this doesn't work yet. 

while the height and width variables are set correctly the URI is not rewritten correctly so the proxy pass lookup fails. 

the default server block requires an extra rewrite rule to match on the none height
```
location / {
            rewrite "^/([1-9][0-9]{0,2}x[1-9][0-9]{0,2})(/.*)$" $2 break;
            rewrite "^/([1-9][0-9]{0,2}xnone)(/.*)$" $2 break;
....
```

Also note: without the location blocks specifiying the url `none` value they won't match and will fallback to the default
> # proxy all other requests to the azure zoonversestatic storage account, $web container

so in theory to support `none` across all hosted domain paths
- panoptes-uploads-staging.zooniverse.org
- panoptes-uploads.zooniverse.org/staging
- panoptes-uploads.zooniverse.org
we need values that match on the none as well (or include the `none` in the current location regexes), e.g. 

```
# Proxy panoptes-uploads production requests to Azure storage
        # e.g. https://panoptes-uploads.zooniverse.org/production/workflow_attached_image/c3303c4d-28b2-453c-882c-63a95a5c41bc.jpeg
        #      https://panoptes-uploads.zooniverse.org/workflow_attached_image/6769554f-9079-41c8-acfc-393737a4972f.jpeg
        location ~ "^/[1-9][0-9]{0,2}xnone/panoptes-uploads.zooniverse.org/.+$" {
            rewrite "^/[1-9][0-9]{0,2}xnone/panoptes-uploads.zooniverse.org(?:/production)?(/.*)$" $1 break;
            resolver 8.8.8.8;
...
# for each current location block**
```

** The default location block (last currently) will be used when no others match so this doesn't need an explicit `none` match.

**Comment by eatyourgreens** [comment]
I've completely rewritten this to make either parameter optional, and update the location blocks to match too. We'll need to update URLs in the front-end, but it now supports URLs like `/300x/` (resize to a fixed width) and `/x300/` (resize to a fixed height.)

I ran it locally with `docker compose up --build` and tested out images in the browser
eg. http://localhost:8080/400x/www.zooniverse.org/291a76c92e4335f7e3a0bed53af6a7bf.jpg

**Comment by eatyourgreens** [comment]
Once this is merged, PFE can be updated to remove the default height of 999 from images.

**Review by camallen — COMMENTED** [review]
This is a big change and it's hard to test. I'd prefer if this change was it's own server block running on `thumbnail-preview.zooniverse.org` in it's own file, i.e. a place to prototype changes for this service. 

We'd have to setup that DNS subdomain to route to the k8s ingress. Also that would require  a k8s ingress mapping to route to this domain.
https://github.com/zooniverse/static/blob/7b4b8cb18c4d8e1fde8b84b184ebdcd565b5459e/kubernetes/ingress/zooniverse.org.yaml#L463-L472
```
host: thumbnails-preview.zooniverse.org
    http:
      paths:
      - pathType: Prefix
        path: /
        backend:
          service:
            name: thumbnailer
            port:
              number: 80
```

I think with a large change like this isolating that change set to a testing sandbox before merging is the best approach.

**Comment by eatyourgreens** [comment]
Ok. Can you set up staging for this repo or should I do it?

I didn’t see any problems when I ran this in Docker but having separate staging and production subdomains for this service would be fine with me. We could run the HEAD of master on staging, and a `production-release` tag on production, like we do with other services.

**Comment by eatyourgreens** [comment]
If we could get this out soonish, that would be great, because I think it will fix the problem of blurry or pixelated thumbnails reported by projects like Notes from Nature.

**Comment by eatyourgreens** [comment]
Living With Machines have an example of the blurry image bug on their home page at the moment. The middle image here is blurred because it’s sized to height, not width, then the width is stretched to fit. 

Anyhow, if we can finally fix this display bug by properly supporting fitting to a fixed width, that would be great. 

Preview images of Victorian newspaper articles from the Living With Machines home page. One is blurred where it has been stretched to fill the available width.

**Comment by camallen** [comment]
> Can you set up staging for this repo or should I do it?

feel free to make a start and i can pick it up tomorrow.

**Comment by eatyourgreens** [comment]
This could be a good opportunity to remove the Jenkinsfile too.

**Comment by eatyourgreens** [comment]
On second thoughts, I'm going to merge this, double check it's working in production and revert if necessary. I haven't been able to break it in a local Docker image, so I'm reasonably confident it works.

**Comment by eatyourgreens** [comment]
The deploy looks good and it's fixed preview thumbnails on project home pages. This has been a longstanding bug, so it's good to have a fix.
https://github.com/zooniverse/Panoptes-Front-End/pull/6351

**Pull request merged**